### PR TITLE
[Reopen][Inductor][CPU] Fuse SmoothQuant int8 linear pattern

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142036

Reopen of https://github.com/pytorch/pytorch/pull/139595

**About the PR**
In the implementation of SmoothQuant in Torchao, quantized linear is computed by `_int_mm(a, b)` + `mul(b_scale)` + `mul(a_scale)` (+ optional `add` for bias) with `reshape` and `convert_dtype` in between.
This PR adds a pass to fuse the corresponding patterns:
- (no bias) `reshape -> _int_mm -> convert_element_type -> (expand -> mul) -> mul -> reshape`
- (with bias) `pattern_no_bias -> add -> reshape -> reshape`

The patterns are replaced by `onednn.qlinear_pointwise` and `onednn.qlinear_prepack`, the latter of which is evaluated and frozen during the freezing process of Inductor. The final graph contains `onednn.qlinear_pointwise` only with packed weight constants.

Note that `onednn.qlinear_pointwise` only supports a scalar activation scale, which is a limitation of oneDNN library, so in that case we set activation scale to 1 and bias to none and apply scales and add bias after `onednn.qlinear_pointwise`.

**Validation results**
Accuracy/perplexity is not changed with or without this fusion pass.
Latency is improved by >10% with the fusion pass.
Test method:
- Model: EleutherAI/gpt-j-6b
- Hardware: Intel(R) Xeon(R) Platinum 8490H, running on 1 socket, 60 cores
- Using Intel OMP and Tcmalloc
- Running [the example script of SmoothQuant in Torchao](https://github.com/pytorch/ao/blob/main/torchao/prototype/smoothquant/example.py) with `TORCHINDUCTOR_FREEZING=1 numactl -N1 python example.py -m EleutherAI/gpt-j-6b --device=cpu --quant-mode=dynamic --compile`

**Test plan**
```
python test/inductor/test_mkldnn_pattern_matcher.py -k test_smooth_quant_with_int_mm
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D66796966](https://our.internmc.facebook.com/intern/diff/D66796966)

### CODE REVIEW

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 CI shows green. Would you like to import it? Thanks.

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66796966).

**Review by jerryzh168 — APPROVED** [review]

**Comment by Xia-Weiwen** [comment]
> ## Merge failed
> **Reason**: This PR has internal changes and must be landed via Phabricator! Please try reimporting/rexporting the PR!
> 
> Details for Dev Infra team

Hi @jerryzh168 It failed to merge by pytorchmergebot. Could you please import it again? Thanks.

**Comment by sanchitintel** [comment]
Sorry for the inconvenience! I guess ghstack also updated this PR's base commit after I submitted a ghstack PR. :(
I won't make any further changes to it until this one's merged.

**Comment by Xia-Weiwen** [comment]
> Sorry for the inconvenience! I guess ghstack also updated this PR's base commit after I submitted a ghstack PR. :( I won't make any further changes to it until this one's merged.

Don't worry. Seems a bug of ghstack. Maybe create an issue to ghstack to talk about this?

**Comment by jerryzh168** [comment]
I'll land from fbcode, but I can't bypass the github failures. please wait until hud (https://hud.pytorch.org/) is green, rebase and ping me again

**Comment by Xia-Weiwen** [comment]
> I'll land from fbcode, but I can't bypass the github failures. please wait until hud (https://hud.pytorch.org/) is green, rebase and ping me again

Hi @jerryzh168 CI checks finished and there are two irrelevant failures. Do I need to rebase? CI will restart after rebase.

**Comment by sanchitintel** [comment]
@Xia-Weiwen, looks like those 2 CI checks are passing in main now. Perhaps we could verify by creating a temporary PR just to run CI after rebasing with main, and then rebase this PR if those CI checks would not fail?

**Comment by Xia-Weiwen** [comment]
> @Xia-Weiwen, looks like those 2 CI checks are passing in main now. Perhaps we could verify by creating a temporary PR just to run CI after rebasing with main, and then rebase this PR if those CI checks would not fail?

Let me rebase this PR.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 The failure looks irrelevant. Could you please take a look and see if it can be imported again? Thanks.

**Comment by sanchitintel** [comment]
@Xia-Weiwen, how about creating a new UT subclass for this PR? 
This PR currently seems to require manual intervention because line 177 on the left does not seem to be present in Meta-internal code -
image

So, if you'd create a new subclass of `TestPatternMatcherBase` and would overload `_test_code_common`, then PR code wouldn't require manual-changes in Meta-internal code.

File: test/inductor/test_mkldnn_pattern_matcher.py
Comment (sanchitintel) [inline test/inductor/test_mkldnn_pattern_matcher.py:179]: @Xia-Weiwen, how about creating a new UT subclass for this PR? 
This PR currently seems to require manual intervention because line 177 on the left does not seem to be present in Meta-internal code -
image

So, if you'd create a new subclass of `TestPatternMatcherBase` and would overload `_test_code_common`, then this PR's code wouldn't require manual-changes in Meta-internal code, and the `Meta Internal-Only Changes Check` CI check will pass.

We might be able to land this PR before v2.6 code freeze with such a change.

Comment (sanchitintel) [reply test/inductor/test_mkldnn_pattern_matcher.py:179]: Sorry, just noticed that `Meta Internal-Only Changes Check` had initially passed with this PR.

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 There is a dependent PR targeting 2.6 branch cut. Could you please check if this PR can be imported and merged? Thanks.

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66796966).

**Review by jgong5 — APPROVED** [review]

**Comment by Xia-Weiwen** [comment]
> Hi @jerryzh168 There is a dependent PR targeting 2.6 branch cut. Could you please check if this PR can be imported and merged? Thanks.

Hi @jerryzh168 We would like this in 2.6. Could you please merge this PR once internal checks show green? Thanks.

**Comment by jerryzh168** [comment]
yeah I'm trying to land this

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D66796966).

**Comment by Xia-Weiwen** [comment]
@jerryzh168 Thanks a lot

**Pull request closed without merging**
